scrapy/core/engine.py

```diff
@@ -165,7 +167,10 @@ def _next_request_from_scheduler(self, spider):
         return d
 
     def _handle_downloader_output(self, response, request, spider):
-        assert isinstance(response, (Request, Response, Failure)), response
+        if not isinstance(response, (Request, Response, Failure)):
+            raise TypeError(
+                "Incorrect type: expected Request, Response or Failure, got %s" % type(response)
```

Comment:
It may be good to also include `repr(response)` in the message (e.g. appending `: %r` to the message and switching `type(response)` to `(type(response), response)`), to make it easier to identify which response triggered the error. Same in similar scenarios in this patch.